src/rtc.c

```diff
@@ -0,0 +1,518 @@
+        a->month == b->month &&
+        a->day   == b->day   &&
+        a->hour  == b->hour  &&
+        a->min   == b->min
+        /* Skip sec comparison in case the clock ticked between RTC reads */
+    );
+}
+
+/**
+ * @brief Read the status of the real-time clock.
+ *
+ * The RTC status response contains an identifier byte to indicate that the
+ * RTC is present on the cartridge and a status byte to indicate whether the
+ * RTC is in the "busy" state.
+ *
+ * Generally, you should not need to call this function directly. Prefer
+ * calling the high-level #joybus_rtc_is_present and #joybus_rtc_is_busy
+ * functions which handle the magic numbers associated with the response.
+ *
+ * @return the RTC status response data
+ */
+uint16_t joybus_rtc_status( void )
+{
+    static const uint64_t input[8] =
+    {
+        0x00000000ff010306,
+        0xfffffffe00000000,
+        0,
+        0,
+        0,
+        0,
+        0,
+        1
+    };
+    static uint64_t output[8];
+
+    controller_exec_pif( input, output );
+
+    uint8_t * recv_bytes = (uint8_t *)&output[1];
+
+    return (uint16_t)recv_bytes[1] << 8 | recv_bytes[2];
+}
+
+/**
+ * @brief Read a block of data from the real-time clock
+ *
+ * @param[in]   block
+ *              Which RTC block to read from (0-2)
+ *
+ * @param[out]  data
+ *              Destination pointer for the RTC block data
+ *
+ * @return the busy status of the real-time clock
+ */
+uint8_t joybus_rtc_read( uint8_t block, uint64_t * data )
+{
+    assert(block <= 2);
+
+    static uint64_t input[8] =
+    {
+        0x0000000002090700,
+        0xffffffffffffffff,
+        0xfffe000000000000,
+        0,
+        0,
+        0,
+        0,
+        1
+    };
+    static uint64_t output[8];
+
+    input[0] = (input[0] & 0xFFFFFFFFFFFFFF00) | block;
+
+    controller_exec_pif( input, output );
+
+    *data = output[1];
+
+    return output[2] >> 56;
+}
+
+/**
+ * @brief Write a block of data to the real-time clock
+ *
+ * @param[in]   block
+ *              Which RTC block to write to (0-2)
+ *
+ * @param[out]  data
+ *              RTC block data to write
+ *
+ * @return the busy status of the real-time clock
+ */
+uint8_t joybus_rtc_write( uint8_t block, const uint64_t * data )
+{
+    assert(block <= 2);
+
+    static uint64_t input[8] =
+    {
+        0x000000000A010800,
+        0xffffffffffffffff,
+        0xfffe000000000000,
+        0,
+        0,
+        0,
+        0,
+        1
+    };
+    static uint64_t output[8];
+
+    input[0] = (input[0] & 0xFFFFFFFFFFFFFF00) | block;
+    input[1] = *data;
+
+    controller_exec_pif( input, output );
+
+    return output[2] >> 56;
+}
+
+/**
+ * @brief Read the status of the real-time clock.
+ *
+ * Some flash carts require the RTC to be explicitly enabled.
+ * Some emulators and flash carts do not support RTC at all.
+ *
+ * @return whether the Joybus RTC is present on the cartridge.
+ */
+bool joybus_rtc_is_present( void )
```

Comment:
Can you please add rough timing in milliseconds to all public function documentation in this file?